bundles/org.vivoweb.dspacevivo.transformation/src/main/java/org/vivoweb/dspacevivo/transformation/harvester/oai/CommunityItr.java

```diff
@@ -0,0 +1,93 @@
+
+public class CommunityItr implements Iterator<Community> {
+
+    private DspaceOAI dspaceHarvester;
+    private List<Community> oaiPage = null;
+    private Item nextCommunity = null;
+
+    public CommunityItr(DspaceOAI hr) {
+        this.dspaceHarvester = hr;
+    }
+
+    @Override
+    public boolean hasNext() {
+        if (oaiPage != null && !oaiPage.isEmpty()) {
+            return true;
+        }
+        OAIPMHResponse oaipmhResponse;
+        boolean hasNext = false;
+        boolean iterate = false;
+        if (!this.dspaceHarvester.isEmpty()) {
+            String responseXML = null;
+            do {
+                try {
+                    responseXML = this.dspaceHarvester.getHttpClient().doRequest(this.dspaceHarvester.getBaseURI(), this.dspaceHarvester.getVerb(), this.dspaceHarvester.getSet(), this.dspaceHarvester.getFrom(), this.dspaceHarvester.getUntil(), this.dspaceHarvester.getMetadata(),
+                            this.dspaceHarvester.getResumptionToken(), this.dspaceHarvester.getIdentifier());
+
+                    oaipmhResponse = new OAIPMHResponse(responseXML, dspaceHarvester.getConf());
+
+                    oaiPage = oaipmhResponse.modelCommunity();
+
+                    HashMap<String, List<Collection>> hmCollection = new HashMap<String, List<Collection>>();
+                    Iterator<Collection> CollectionItr = this.dspaceHarvester.harvestCollection();
+                    while (CollectionItr.hasNext()) {
+                        Collection cl = CollectionItr.next();
+                        for (String comId : cl.getIsPartOfCommunityID()) {
+                            if (!hmCollection.containsKey(comId)) {
+                                hmCollection.put(comId, new ArrayList());
+
+                            }
+                            hmCollection.get(comId).add(cl);
+                        }
+
+                    }
+
+                    for (Community com : oaiPage) {
+                        String id = com.getId();
+                        com.setHasCollection(hmCollection.get(id));
+
+                    }
+                } catch (Exception ex) {
+                    return hasNext;
+                }
+
+                Optional<String> resumptionToken = oaipmhResponse.getResumptionToken();
+                if (resumptionToken.isPresent() && !resumptionToken.get().isEmpty()) {
+                    this.dspaceHarvester.setResumptionToken(resumptionToken.get());
+                } else {
+                    this.dspaceHarvester.setResumptionToken(null);
+                    this.dspaceHarvester.setEmpty(true);
+
+                }
+
+                if (!oaiPage.isEmpty()) {
+                    hasNext = true;
+                } else if (this.dspaceHarvester.getResumptionToken() != null) {
+                    iterate = true;
+                }
+
+            } while (iterate);
+
+        }
+        return hasNext;
+    }
+
+    @Override
+    public Community next() {
```

Comment:
Again, I don't like how this is implemented.

Reply:
To get the data from Dspace multiple OAI requests have to be performed. Each OAI response returns a list (page) of elements e.g.  Items, Communities, Collections. In these Iterator classes I am converting the responses from pages to a stream of elements, so the complexity of pagination in the OAI requests are hidden behind the Iterator class. this is the reason why I am using a temporary list to store the results of the current page and in the hasNext method the OAI requests are executed when the temporary list is empty. Please let me know if you have suggestions on how to improve it